File: apps/server/internal/protocol/protocol.go
package protocol

import (
	"encoding/json"
	"errors"
)

const (
	TypeJoinRoom     = "join_room"
	TypeMakeMove     = "make_move"
	TypeRoomState    = "room_state"
	TypePlayerJoined = "player_joined"
	TypePlayerLeft   = "player_left"
	TypeMoveApplied  = "move_applied"
	TypeMoveRejected = "move_rejected"
	TypeError        = "error"
)

type Seat string

const (
	SeatPlayer1  Seat = "player1"
	SeatPlayer2  Seat = "player2"
	SeatObserver Seat = "observer"
)

type Envelope struct {
	Type    string          `json:"type"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

type JoinRoom struct {
	RoomCode string `json:"roomCode"`
}

type ClientMove struct {
	Seq       int    `json:"seq"`
	Kind      string `json:"kind,omitempty"`
	Orig      string `json:"orig,omitempty"`
	Dest      string `json:"dest,omitempty"`
	Promotion string `json:"promotion,omitempty"`
	Color     string `json:"color,omitempty"`
}

type Move struct {
	Seq       int    `json:"seq"`
	Seat      Seat   `json:"seat"`
	Kind      string `json:"kind,omitempty"`
	Orig      string `json:"orig,omitempty"`
	Dest      string `json:"dest,omitempty"`
	Promotion string `json:"promotion,omitempty"`
	Color     string `json:"color,omitempty"`
}

type Player struct {
	ID   string `json:"id"`
	Seat Seat   `json:"seat"`
}

type RoomState struct {
	RoomCode string   `json:"roomCode"`
	You      Player   `json:"you"`
	Players  []Player `json:"players"`
	Turn     Seat     `json:"turn"`
	Seq      int      `json:"seq"`
	Moves    []Move   `json:"moves"`
}

type PlayerEvent struct {
	Player Player `json:"player"`
}

type MoveRejected struct {
	Reason string `json:"reason"`
	Seq    int    `json:"seq"`
}

type Error struct {
	Message string `json:"message"`
}

func Decode[T any](payload json.RawMessage) (T, error) {
	var value T
	if len(payload) == 0 {
		return value, errors.New("missing payload")
	}
	if err := json.Unmarshal(payload, &value); err != nil {
		return value, err
	}
	return value, nil
}

func Encode(messageType string, payload any) ([]byte, error) {
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}
	return json.Marshal(Envelope{Type: messageType, Payload: body})
}

File: apps/web/src/game/online/protocol.ts
import type * as types from "../rules/types.ts";
import type { PieceColor, Role } from "@osc/rules";

export type Seat = "player1" | "player2" | "observer";

export interface Player {
  id: string;
  seat: Seat;
}

export interface Move {
  seq: number;
  seat: Seat;
  kind?: "move";
  orig: types.Key;
  dest: types.Key;
  promotion?: Role;
}

export interface CastleMove {
  seq: number;
  seat: Seat;
  kind: "castle";
  orig: types.Key;
  dest: types.Key;
}

export interface DefectionMove {
  seq: number;
  seat: Seat;
  kind: "defect";
  color: PieceColor;
}

export type RoomMove = Move | CastleMove | DefectionMove;

export interface RoomState {
  roomCode: string;
  you: Player;
  players: Player[];
  turn: Seat;
  seq: number;
  moves: RoomMove[] | null;
}

export interface MoveRejected {
  reason: string;
  seq: number;
}

export interface ServerError {
  message: string;
}

export type ServerMessage =
  | { type: "room_state"; payload: RoomState }
  | { type: "player_joined"; payload: { player: Player } }
  | { type: "player_left"; payload: { player: Player } }
  | { type: "move_applied"; payload: RoomMove }
  | { type: "move_rejected"; payload: MoveRejected }
  | { type: "error"; payload: ServerError };

export interface ClientMove {
  seq: number;
  kind?: "move";
  orig: types.Key;
  dest: types.Key;
  promotion?: Role;
}

export interface ClientCastle {
  seq: number;
  kind: "castle";
  orig: types.Key;
  dest: types.Key;
}

export interface ClientDefection {
  seq: number;
  kind: "defect";
  color: PieceColor;
}

export type ClientAction = ClientMove | ClientCastle | ClientDefection;

export function makeMoveMessage(move: ClientAction): string {
  return JSON.stringify({ type: "make_move", payload: move });
}

export function normalizeRoomCode(value: string | undefined | null): string {
  return (value ?? "").trim().toUpperCase();
}

export function roomCodeFromLocation(location: Location): string {
  return normalizeRoomCode(new URL(location.href).searchParams.get("room"));
}

export function playRoomUrl(
  roomCode: string,
  origin = window.location.origin,
): string {
  const url = new URL("/play/", origin);
  url.searchParams.set("room", normalizeRoomCode(roomCode));
  return url.toString();
}

export function serverHttpOrigin(): string {
  const configured = import.meta.env.PUBLIC_OSC_SERVER_URL;
  if (configured) {
    return configured;
  }
  if (
    window.location.hostname === "127.0.0.1" ||
    window.location.hostname === "localhost"
  ) {
    return `${window.location.protocol}//${window.location.hostname}:8080`;
  }
  return window.location.origin;
}

export function createRoomEndpoint(origin = serverHttpOrigin()): string {
  return new URL("/api/rooms", origin).toString();
}

export function roomWebsocketUrl(
  roomCode: string,
  origin = serverHttpOrigin(),
): string {
  const url = new URL(`/api/rooms/${normalizeRoomCode(roomCode)}/ws`, origin);
  url.protocol = url.protocol === "https:" ? "wss:" : "ws:";
  return url.toString();
}
